### PR TITLE
BaSP-TG-115: Project form update

### PR DESCRIPTION
**Pull request by MaicolFreire001**
**What's have been done:**
- Replaced the actual logic for react hook form logic
- Added joi validations and error messages

**How to test:**
Check if the project entity works properly, all inputs have the corrects validations and the errors are displayed correctly
**Important note:** The role field doesn't show its error but it works fine, this issue will be resolved later.

### CODE REVIEW

File: src/Components/Projects/Form/index.js [hunk 1]
```diff
@@ -57,31 +56,44 @@ const Project = () => {
...
-            employee: newProjectList[0].employees[0].employee._id,
-            role: newProjectList[0].employees[0].role,
-            rate: newProjectList[0].employees[0].rate
-          }
-        ]
-      });
-    } else {
-      setIsFetched(false);
+      if (projectData === undefined) {
+        fetch(`${process.env.REACT_APP_API_URL}/projects/${projectId}`)
```
Comment (ignacioaristo) [inline src/Components/Projects/Form/index.js:79]: If we are goin to handle an async external call, we should handle it into the thunks.

**Review by ignacioaristo — DISMISSED** [review]

File: src/Components/Projects/Form/index.js (hunk 1 above)
Comment (MaicolFreire001) [reply src/Components/Projects/Form/index.js]: Done

File: src/Components/Projects/Form/index.js [hunk 2]
```diff
@@ -100,19 +129,10 @@ const Project = () => {
...
-    } else {
-      setProjectBody({ ...projectBody, [key]: value });
-    }
-  };
-
-  const onSubmit = () => {
-    projectId ? dispatch(putProject(projectId, projectBody)) : dispatch(postProject(projectBody));
+  const onSubmit = (data) => {
+    console.log('Test');
+    console.log(data);
```
Comment (JuliMonte5) [inline src/Components/Projects/Form/index.js:134]: Remove these console logs

**Review by JuliMonte5 — CHANGES_REQUESTED** [review]

File: src/Components/Projects/Form/index.js (hunk 2 above)
Comment (MaicolFreire001) [reply src/Components/Projects/Form/index.js]: Done

File: src/Components/Projects/validations.jsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+export const schema = Joi.object({
+  name: Joi.string()
+    .min(3)
+    .trim()
+    .regex(/^(?=.*[a-zA-Z].*)([\w\s\W]+)$/)
+    .required()
+    .messages({
+      'string.empty': 'Project name is required',
+      'string.pattern.base': 'Project name should have letters only',
```
Comment (fede230298) [inline src/Components/Projects/validations.jsx:21]: This regex does not mean this, the regex condition is to have at least one letter. It accepts numbers and special chars. But requires at least one letter out of the minimum three characters required

File: src/Components/Projects/validations.jsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      'any.required': 'Project name is required'
+    }),
+  description: Joi.string()
+    .min(5)
+    .trim()
+    .regex(/^(?=.*[a-zA-Z].*)([\w\s\W]+)$/)
+    .required()
+    .messages({
+      'string.empty': 'Description is required',
+      'string.pattern.base': 'Name should have letters only',
```
Comment (fede230298) [inline src/Components/Projects/validations.jsx:32]: Same as above, it means it requires at least one letter

File: src/Components/Projects/validations.jsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    'any.required': 'Start date is required'
+  }),
+  endDate: Joi.date().greater(Joi.ref('startDate')).messages({
+    'date.greater': 'End date must be newer than start date',
+    'any.required': 'End date is required'
+  }),
+  clientName: Joi.string()
+    .min(3)
+    .trim()
+    .regex(/^(?=.*[a-zA-Z].*)([\w\s\W]+)$/)
```
Comment (fede230298) [inline src/Components/Projects/validations.jsx:46]: Same as above, it means it requires at least one letter

File: src/Components/Projects/validations.jsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    .required()
+    .messages({
+      'string.empty': 'Client name is required',
+      'string.pattern.base': 'Client name should have letters only',
+      'string.min': 'Client name should have a minimum length of 2 characters',
```
Comment (fede230298) [inline src/Components/Projects/validations.jsx:51]: Minimum length of 3 characters

File: src/Components/Projects/validations.jsx (hunk 3 above)
Comment (MaicolFreire001) [reply src/Components/Projects/validations.jsx:21]: Solved

File: src/Components/Projects/validations.jsx (hunk 4 above)
Comment (MaicolFreire001) [reply src/Components/Projects/validations.jsx:32]: Solved

File: src/Components/Projects/validations.jsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    'any.required': 'End date is required'
+  }),
+  clientName: Joi.string()
+    .min(3)
+    .trim()
+    .regex(/^(?=.*[a-zA-Z].*)([\w\s\W]+)$/)
+    .required()
+    .messages({
+      'string.empty': 'Client name is required',
+      'string.pattern.base': 'Client name should have letters only',
```
[the remark this answers is not in the document]
Comment (MaicolFreire001) [reply src/Components/Projects/validations.jsx:50]: Solved

File: src/Components/Projects/validations.jsx (hunk 6 above)
Comment (MaicolFreire001) [reply src/Components/Projects/validations.jsx:51]: Solved

**Review by JuliMonte5 — APPROVED** [review]

**Review by GonzaloAsencio — APPROVED** [review]

**Review by ProDIGGY-vinilos — APPROVED** [review]

**Review by fede230298 — APPROVED** [review]

**Review by sofihoj — APPROVED** [review]

File: src/Components/Projects/Form/index.js [hunk 8]
```diff
@@ -100,19 +131,10 @@ const Project = () => {
...
-      });
-    } else {
-      setProjectBody({ ...projectBody, [key]: value });
-    }
-  };
-
-  const onSubmit = () => {
-    projectId ? dispatch(putProject(projectId, projectBody)) : dispatch(postProject(projectBody));
+  const onSubmit = (data) => {
+    console.log(errors.endDate);
```
[the remark this answers is not in the document]
Comment (MaicolFreire001) [reply src/Components/Projects/Form/index.js]: Done

**Pull request merged**